flows/engine/session.go

```diff
@@ -157,8 +157,16 @@ func (s *session) Resume(callerEvents []flows.Event) error {
 
 	// events can change run status so only proceed to the wait if we're still waiting
 	if waitingRun.Status() == flows.RunStatusWaiting {
-		if s.wait.CanResume(waitingRun, step) {
-			waitingRun.SetStatus(flows.RunStatusActive)
+		waitCanResume := s.wait.CanResume(waitingRun, step)
+		waitHasTimedOut := s.wait.HasTimedOut()
+
+		if waitCanResume || waitHasTimedOut {
+			if waitCanResume {
+				s.wait.Resume(waitingRun)
+			} else {
+				s.wait.ResumeByTimeOut(waitingRun)
```

Comment:
I think only a msg wait should clear the input if it times out - so making waits handle timing out themselves.